src/quickgui/qgsquickhighlightsgnode.cpp

```diff
@@ -0,0 +1,167 @@
+    {
+      QgsGeometry geomN( collection->geometryN( i )->clone() );
+      handleSingleGeometry( geomN );
+    }
+  }
+  else
+  {
+    handleSingleGeometry( geom );
+  }
+}
+
+void QgsQuickHighlightSGNode::handleSingleGeometry( const QgsGeometry &geom )
+{
+  Q_ASSERT( !geom.isMultipart() );
+
+  switch ( geom.type() )
+  {
+    case QgsWkbTypes::PointGeometry:
+    {
+      QVector<QgsPoint> points;
+      for ( auto it = geom.vertices_begin(); it != geom.vertices_end(); ++it )
+        points.append( *it );
+
+      if ( !points.isEmpty() )
+        appendChildNode( createPointGeometry( points.at( 0 ) ) );
+      break;
+    }
+
+    case QgsWkbTypes::LineGeometry:
+    {
+      QVector<QgsPoint> points;
+      for ( auto it = geom.vertices_begin(); it != geom.vertices_end(); ++it )
+        points.append( *it );
+
+      appendChildNode( createLineGeometry( points ) );
+      break;
+    }
+
+    case QgsWkbTypes::PolygonGeometry:
+    {
+      const QgsPolygon *poly = qgsgeometry_cast<const QgsPolygon *>( geom.constGet() );
+      if ( poly )
+        appendChildNode( createPolygonGeometry( *poly ) );
+      break;
+    }
+
+    case QgsWkbTypes::UnknownGeometry:
+    case QgsWkbTypes::NullGeometry:
+      break;
+  }
+}
+
+QSGGeometryNode *QgsQuickHighlightSGNode::createLineGeometry( const QVector<QgsPoint> &points )
+{
+  std::unique_ptr<QSGGeometryNode> node = qgis::make_unique< QSGGeometryNode>();
+  std::unique_ptr<QSGGeometry> sgGeom = qgis::make_unique< QSGGeometry>( QSGGeometry::defaultAttributes_Point2D(), points.count() );
+  QSGGeometry::Point2D *vertices = sgGeom->vertexDataAsPoint2D();
+
+  int i = 0;
+  for ( const QgsPoint &pt : points )
+  {
+    vertices[i++].set(
+      static_cast< float >( pt.x() ),
+      static_cast< float >( pt.y() )
+    );
+  }
+
+  sgGeom->setLineWidth( mWidth );
+  sgGeom->setDrawingMode( GL_LINE_STRIP );
+  node->setGeometry( sgGeom.release() );
+  node->setMaterial( &mMaterial );
+  node->setFlag( QSGNode::OwnsGeometry );
+  node->setFlag( QSGNode::OwnedByParent );
+  return node.release();
+}
+
+QSGGeometryNode *QgsQuickHighlightSGNode::createPointGeometry( const QgsPoint &point )
+{
+  std::unique_ptr<QSGGeometryNode> node = qgis::make_unique< QSGGeometryNode>();
+  std::unique_ptr<QSGGeometry> sgGeom = qgis::make_unique<QSGGeometry>( QSGGeometry::defaultAttributes_Point2D(), 1 );
+
+  QSGGeometry::Point2D *vertices = sgGeom->vertexDataAsPoint2D();
+  vertices[0].set(
+    static_cast< float >( point.x() ),
+    static_cast< float >( point.y() )
+  );
+  sgGeom->setDrawingMode( GL_POINTS );
+  sgGeom->setLineWidth( mWidth );
+
+  node->setGeometry( sgGeom.release() );
+  node->setMaterial( &mMaterial );
+  node->setFlag( QSGNode::OwnsGeometry );
+  node->setFlag( QSGNode::OwnedByParent );
+  return node.release();
+}
+
+QSGGeometryNode *QgsQuickHighlightSGNode::createPolygonGeometry( const QgsPolygon &polygon )
+{
+  QgsTessellator tes( 0.0, 0.0, false, false, false );
```

Comment:
would be good to set origin based on polygon's first vertex or so, because vertices are stored as floats and we would be loosing precision. It should be possible to set offset transform to the QSGGeometryNode to shift the origin again.